### PR TITLE
feat: add gateway topology with endpoint visibility support

### PR DESCRIPTION
**Pull request by NomadXD**
## Purpose
This PR adds support for gateway topology at the dataplane and environment level and support endpoint URL resolution for endpoint visibilities.

## Approach
- Introduce a new GatewaySpec schema to represent the gateway topology at the dataplane and environment level
- Add a new API to the CEL context named `gateway` where the platform engineers can use to access the gateway without explicitly defining the gateway topology
- Continue to support `dataplane.gateway` and `environment.gateway` APIs for other edge cases or advanced use cases
- Reimplement endpoint URL resolution based on endpoint visibility and type

## Related Issues
https://github.com/openchoreo/openchoreo/issues/1797

## Checklist
- [x] Tests added or updated (unit, integration, etc.)
- [ ] Samples updated (if applicable)

## Remarks
Samples related to this will be updated separately with https://github.com/openchoreo/openchoreo/issues/2190

### CODE REVIEW

File: internal/controller/releasebinding/controller.go [hunk 1]
```diff
@@ -1019,91 +1045,105 @@ func resolveEndpointURLStatuses(
...
+			routeIndex[endpointName] = &endpointRoutes{}
+		}
+
+		visibility := openchoreov1alpha1.EndpointVisibility(objLabels[labels.LabelKeyEndpointVisibility])
+		if visibility == openchoreov1alpha1.EndpointVisibilityExternal {
+			routeIndex[endpointName].external = obj
+		} else {
+			routeIndex[endpointName].internal = obj
+		}
+	}
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/controller/releasebinding/controller.go:1070]: Only external and internal visibilities will generate HTTPRoutes, hence this is not required to handle.

File: internal/controller/releasebinding/controller.go [hunk 2]
```diff
@@ -1142,73 +1182,36 @@ func extractFirstPathValue(obj *unstructured.Unstructured) string {
...
-		if spec.OrganizationHTTPSPort != 0 {
-			return spec.OrganizationHTTPSPort
-		}
-		return standardHTTPSPort
+	if visibility == openchoreov1alpha1.EndpointVisibilityExternal {
+		return spec.Ingress.External
 	}
-	return 0
+	return spec.Ingress.Internal
 }
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/controller/releasebinding/controller.go:1212]: For the initial version and simplicity, this behaviour is okay for now. We'll incrementally improve the gateway topology resolution logic further with different dimensions like internal/external and ingress/egress.

File: install/helm/openchoreo-control-plane/crds/openchoreo.dev_environments.yaml [hunk 3]
```diff
@@ -64,54 +64,302 @@ spec:
...
+                                format: int32
+                                maximum: 65535
+                                minimum: 1
+                                type: integer
+                            required:
+                            - host
+                            - port
+                            type: object
+                        type: object
+                    type: object
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply install/helm/openchoreo-control-plane/crds/openchoreo.dev_environments.yaml:362]: breaking changes are expected. will update the release note with migration plan

File: internal/controller/releasebinding/controller.go [hunk 4]
```diff
@@ -1019,91 +1042,105 @@ func resolveEndpointURLStatuses(
...
 	}
-
-	firstRef, ok := backendRefs[0].(map[string]interface{})
-	if !ok {
-		return 0
+	if ep.TLS != nil {
+		urls.TLS = buildInvokeURL("https", hostname, path, ep.TLS.Port)
 	}
+	return urls
+}
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/controller/releasebinding/controller.go:1134]: This is expected as we cannot derive a port when the listeners are optional.
